das/daser.go

```diff
@@ -68,268 +90,89 @@ func (d *DASer) Start(context.Context) error {
-				d.state.catchUpLk.Unlock()
-
-				log.Errorw("catch-up routine failed", "attempted range: from", job.from, "to", job.to)
-				log.Warn("DASer CATCH-UP SAMPLING ROUTINE WILL BE STOPPED. IN ORDER TO CONTINUE SAMPLING, " +
-					"RE-START THE NODE")
-				return
-			}
-		}
+	if err = d.store.store(ctx, cp); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
 	}
-}
 
-// catchUp starts a sampling routine for headers starting at the next header
-// after the `from` height and exits the loop once `to` is reached. (from:to]
-func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
-	log.Infow("sampling past headers", "from", job.from, "to", job.to)
-
-	// start sampling from height at checkpoint+1 since the
-	// checkpoint height is DASed by broader sample routine
-	for height := job.from + 1; height <= job.to; height++ {
-		h, err := d.getter.GetByHeight(ctx, uint64(height))
-		if err != nil {
-			if err == context.Canceled {
-				// report previous height as the last successfully sampled height and
-				// error as nil since the routine was ordered to stop
-				return height - 1, nil
-			}
+	d.cancel()
+	if err = d.sampler.wait(ctx); err != nil {
+		return fmt.Errorf("DASer force quit: %w", err)
+	}
 
-			log.Errorw("failed to get next header", "height", height, "err", err)
-			// report previous height as the last successfully sampled height
-			return height - 1, err
-		}
+	// save updated checkpoint after sampler and all workers are shut down
+	if err = d.store.store(ctx, newCheckpoint(d.sampler.state.unsafeStats())); err != nil {
+		log.Errorw("storing checkpoint to disk", "Err", err)
+	}
 
-		err = d.sampleHeader(ctx, h)
-		if err != nil {
-			return h.Height - 1, err
-		}
-		d.state.catchUpLk.Lock()
-		d.state.catchUp.Height = uint64(h.Height)
-		d.state.catchUpLk.Unlock()
+	if err = d.store.wait(ctx); err != nil {
+		return fmt.Errorf("DASer force quit with err: %w", err)
 	}
-	d.state.catchUpLk.Lock()
-	d.state.catchUp.End = time.Now()
-	d.state.catchUpLk.Unlock()
-
-	jobDetails := d.CatchUpRoutineState()
-	log.Infow("successfully sampled past headers", "from", job.from,
-		"to", job.to, "finished (s)", jobDetails.Duration())
-	// report successful result
-	return job.to, nil
+	return d.subscriber.wait(ctx)
 }
 
-func (d *DASer) sampleHeader(ctx context.Context, h *header.ExtendedHeader) error {
-	startTime := time.Now()
-
+func (d *DASer) sample(ctx context.Context, h *header.ExtendedHeader) error {
 	err := d.da.SharesAvailable(ctx, h.DAH)
 	if err != nil {
 		if err == context.Canceled {
-			return nil
+			return err
 		}
 		var byzantineErr *ipld.ErrByzantine
 		if errors.As(err, &byzantineErr) {
 			log.Warn("Propagating proof...")
 			sendErr := d.bcast.Broadcast(ctx, fraud.CreateBadEncodingProof(h.Hash(), uint64(h.Height), byzantineErr))
 			if sendErr != nil {
-				log.Errorw("fraud proof propagating failed", "err", sendErr)
+				log.Errorw("fraud proof propagating failed", "Err", sendErr)
```

Comment:
```suggestion
				log.Errorw("fraud proof propagating failed", "err", sendErr)
```

Is this a typo or did you intentionally make it capitalized?